_compdemos/python_logging_demo.md

```diff
@@ -0,0 +1,236 @@
+of different possible output sources. Some of the more common handlers include `StreamHandlers` for writing to various output 
+streams such as STDOUT and STDERR, and `FileHandlers` for writing records directly to a file. It is common to want to see the output of your 
+code in the terminal when you run it, but you might also want output logged to a file somewhere in case your project is run on a cron or run 
+by other remote users. There are a lot of different types of handlers, including email, socket handlers, rotating file handlers, HTTP handlers, and more but I'm not going into them here.
+
+Different log handler classes can be found listed [here](https://docs.python.org/3.8/library/logging.handlers.html#module-logging.handlers)
+
+### Formatters
+`Formatter` objects determine the structure that every message will be reported in, and allow you to add contextual information
+to every log message, such as line number, module name, function name, and date / time. The list of contextual information
+you can add to your message by embedding the key words into your Formatter string can be found 
+[here](https://docs.python.org/3.8/library/logging.html#logging.LogRecord).  Some of the attributes I find myself using
+quite often are `%(lineno)d` (line number), `%(asctime)s` (timestamp), `%(filename)`, `(funcName)s`, and `%(module)s`.  
+There are also some handy formatting attributes for reporting what thread or process a message is coming from if you're 
+working in a concurrent environment.
+
+When creating a Formatter object, it simply must be added to a Handler class instance using the `.setFormatter()` function,
+as seen in the earlier example.  By attaching different `Formatter` objects to different handlers, you can provide different levels
+of detail to different outputs.  For instance maybe I don't want the module, function, and file name cluttering up
+my console, so I could attach a formatter to it which doesn't have those attributes.  But I want my file handler to have as
+much information as possible so I can better reconstruct events when doing retrospectives (example below).
+
+```python
+import logging
+logger = logging.getLogger('new_logger')
+logger.setLevel(logging.DEBUG)
+file_formatter = logging.Formatter('%(asctime)s - %(levelname)s - %(message)s - line %(lineno)s - module %(module)s - function %(funcName)s - file %(fileName)s')
+file_handler = logging.FileHandler(filename='verbose_logger.log')
+file_handler.setFormatter(file_formatter)
+
+slim_formatter = logging.Formatter('%(asctime)s - %(levelname)s - %(message)s - line %(lineno)s')
+slim_console_handler = logging.StreamHandler()
+slim_console_handler.setFormatter(slim_formatter)
+logger.addHandler(slim_console_handler)
+logger.addHandler(file_handler)
+```
+
+### Filters 
+`Filters` are attached to `Handlers` and provided filtering of messages from below a specified point in the logger hierarchy 
```

Comment:
this presently isn't shown above in the code snippet: the filter is being applied at the logger level-- perhaps a second code snippet adding two FileHandlers: infoLogs and errorLogs with different message levels specified?